File: src-tauri/src/commands/vector_db.rs
use crate::state::AppState;
use log::{debug, error, info, warn};
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::collections::HashMap;
use tauri::{State, Window};

#[derive(Debug, Serialize, Deserialize)]
pub struct SearchResultResponse {
    pub id: String,
    pub score: f32,
    pub payload: HashMap<String, Value>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct VectorDbStats {
    pub collection_info: Value,
    pub point_count: u64,
    pub mapped_files: usize,
    pub embedding_model: String,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct AddFileResponse {
    pub point_id: String,
    pub file_path: String,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct SyncFilesResponse {
    pub added_files: Vec<String>,
    pub count: usize,
}

#[tauri::command]
pub async fn create_vector_db(state: State<'_, AppState>, name: String) -> Result<(), String> {
    #[cfg(feature = "vector_db")]
    {
        let mut db_guard = state.vector_db.lock().await;
        if db_guard.is_some() {
            return Err("Vector database already exists".to_string());
        }

        // Create new vector database with the given name as collection
        let embed_config = file_vec::create_default_config()
            .map_err(|e| format!("Failed to create embed config: {}", e))?;

        let mut qdrant_config = embed_config.qdrant_config;
        qdrant_config.collection_name = name.clone();

        match file_vec::VectorDb::new(embed_config.embed_config, qdrant_config).await {
            Ok(db) => {
                *db_guard = Some(db);
                info!("Created vector database with collection: {}", name.clone());
                Ok(())
            }
            Err(e) => Err(format!("Failed to create vector database: {}", e)),
        }
    }
    #[cfg(not(feature = "vector_db"))]
    {
        Err("Vector database feature not enabled".to_string())
    }
}

#[tauri::command]
pub async fn delete_vector_db(state: State<'_, AppState>) -> Result<(), String> {
    #[cfg(feature = "vector_db")]
    {
        let mut db_guard = state.vector_db.lock().await;
        if let Some(db) = db_guard.as_ref() {
            db.delete_all_embeddings()
                .await
                .map_err(|e| format!("Failed to clear vector database: {}", e))?;
        }
        *db_guard = None;
        info!("Deleted vector database");
        Ok(())
    }
    #[cfg(not(feature = "vector_db"))]
    {
        Err("Vector database feature not enabled".to_string())
    }
}

#[tauri::command]
pub async fn add_file_to_vector_db(
    state: State<'_, AppState>,
    file_path: String,
) -> Result<AddFileResponse, String> {
    #[cfg(feature = "vector_db")]
    {
        let db_guard = state.vector_db.lock().await;
        if let Some(ref db) = *db_guard {
            match db.add_file(&file_path).await {
                Ok(point_id) => {
                    info!("Added file '{}' to vector database", file_path);
                    Ok(AddFileResponse {
                        point_id,
                        file_path,
                    })
                }
                Err(e) => Err(format!("Failed to add file to vector database: {}", e)),
            }
        } else {
            Err("Vector database not initialized".to_string())
        }
    }
    #[cfg(not(feature = "vector_db"))]
    {
        Err("Vector database feature not enabled".to_string())
    }
}

#[tauri::command]
pub async fn search_vector_db(
    state: State<'_, AppState>,
    query: String,
    top_k: Option<usize>,
) -> Result<Vec<SearchResultResponse>, String> {
    #[cfg(feature = "vector_db")]
    {
        let db_guard = state.vector_db.lock().await;
        if let Some(ref db) = *db_guard {
            let limit = top_k.unwrap_or(10);
            match db.search(&query, limit).await {
                Ok(results) => {
                    let response: Vec<SearchResultResponse> = results
                        .into_iter()
                        .map(|r| SearchResultResponse {
                            id: r.id.to_string(),
                            score: r.score,
                            payload: r.payload,
                        })
                        .collect();

                    debug!(
                        "Search query '{}' returned {} results",
                        query,
                        response.len()
                    );
                    Ok(response)
                }
                Err(e) => Err(format!("Failed to search vector database: {}", e)),
            }
        } else {
            Err("Vector database not initialized".to_string())
        }
    }
    #[cfg(not(feature = "vector_db"))]
    {
        Err("Vector database feature not enabled".to_string())
    }
}

#[tauri::command]
pub async fn search_vector_db_with_filter(
    state: State<'_, AppState>,
    query: String,
    top_k: Option<usize>,
    filter: HashMap<String, Value>,
) -> Result<Vec<SearchResultResponse>, String> {
    #[cfg(feature = "vector_db")]
    {
        let db_guard = state.vector_db.lock().await;
        if let Some(ref db) = *db_guard {
            let limit = top_k.unwrap_or(10);
            match db.search_with_filter(&query, limit, filter).await {
                Ok(results) => {
                    let response: Vec<SearchResultResponse> = results
                        .into_iter()
                        .map(|r| SearchResultResponse {
                            id: r.id.to_string(),
                            score: r.score,
                            payload: r.payload,
                        })
                        .collect();

                    debug!(
                        "Filtered search query '{}' returned {} results",
                        query,
                        response.len()
                    );
                    Ok(response)
                }
                Err(e) => Err(format!(
                    "Failed to search vector database with filter: {}",
                    e
                )),
            }
        } else {
            Err("Vector database not initialized".to_string())
        }
    }
    #[cfg(not(feature = "vector_db"))]
    {
        Err("Vector database feature not enabled".to_string())
    }
}

#[tauri::command]
pub async fn delete_file_from_vector_db(
    state: State<'_, AppState>,
    file_path: String,
) -> Result<(), String> {
    #[cfg(feature = "vector_db")]
    {
        let db_guard = state.vector_db.lock().await;
        if let Some(ref db) = *db_guard {
            match db.delete_file(&file_path).await {
                Ok(_) => {
                    info!("Deleted file '{}' from vector database", file_path);
                    Ok(())
                }
                Err(e) => Err(format!("Failed to delete file from vector database: {}", e)),
            }
        } else {
            Err("Vector database not initialized".to_string())
        }
    }
    #[cfg(not(feature = "vector_db"))]
    {
        Err("Vector database feature not enabled".to_string())
    }
}

#[tauri::command]
pub async fn delete_point_from_vector_db(
    state: State<'_, AppState>,
    point_id: String,
) -> Result<(), String> {
    #[cfg(feature = "vector_db")]
    {
        let db_guard = state.vector_db.lock().await;
        if let Some(ref db) = *db_guard {
            let uuid = uuid::Uuid::parse_str(&point_id)
                .map_err(|e| format!("Invalid UUID format: {}", e))?;

            match db.delete_point(&uuid).await {
                Ok(_) => {
                    info!("Deleted point '{}' from vector database", point_id);
                    Ok(())
                }
                Err(e) => Err(format!(
                    "Failed to delete point from vector database: {}",
                    e
                )),
            }
        } else {
            Err("Vector database not initialized".to_string())
        }
    }
    #[cfg(not(feature = "vector_db"))]
    {
        Err("Vector database feature not enabled".to_string())
    }
}

#[tauri::command]
pub async fn delete_all_embeddings(state: State<'_, AppState>) -> Result<(), String> {
    #[cfg(feature = "vector_db")]
    {
        let db_guard = state.vector_db.lock().await;
        if let Some(ref db) = *db_guard {
            match db.delete_all_embeddings().await {
                Ok(_) => {
                    info!("Deleted all embeddings from vector database");
                    Ok(())
                }
                Err(e) => Err(format!("Failed to delete all embeddings: {}", e)),
            }
        } else {
            Err("Vector database not initialized".to_string())
        }
    }
    #[cfg(not(feature = "vector_db"))]
    {
        Err("Vector database feature not enabled".to_string())
    }
}

#[tauri::command]
pub async fn get_vector_db_stats(state: State<'_, AppState>) -> Result<VectorDbStats, String> {
    #[cfg(feature = "vector_db")]
    {
        let db_guard = state.vector_db.lock().await;
        if let Some(ref db) = *db_guard {
            match db.get_stats().await {
                Ok(stats_value) => {
                    let stats = VectorDbStats {
                        collection_info: stats_value["collection_info"].clone(),
                        point_count: stats_value["point_count"].as_u64().unwrap_or(0),
                        mapped_files: stats_value["mapped_files"].as_u64().unwrap_or(0) as usize,
                        embedding_model: stats_value["embedding_model"]
                            .as_str()
                            .unwrap_or("unknown")
                            .to_string(),
                    };
                    Ok(stats)
                }
                Err(e) => Err(format!("Failed to get vector database stats: {}", e)),
            }
        } else {
            Err("Vector database not initialized".to_string())
        }
    }
    #[cfg(not(feature = "vector_db"))]
    {
        Err("Vector database feature not enabled".to_string())
    }
}

#[tauri::command]
pub async fn sync_files_to_vector_db(
    state: State<'_, AppState>,
    directory_path: String,
) -> Result<SyncFilesResponse, String> {
    #[cfg(feature = "vector_db")]
    {
        let db_guard = state.vector_db.lock().await;
        if let Some(ref db) = *db_guard {
            match db.sync_files(&directory_path).await {
                Ok(added_files) => {
                    let count = added_files.len();
                    info!(
                        "Synced {} files to vector database from '{}'",
                        count, directory_path
                    );
                    Ok(SyncFilesResponse { added_files, count })
                }
                Err(e) => Err(format!("Failed to sync files to vector database: {}", e)),
            }
        } else {
            Err("Vector database not initialized".to_string())
        }
    }
    #[cfg(not(feature = "vector_db"))]
    {
        Err("Vector database feature not enabled".to_string())
    }
}

#[tauri::command]
pub async fn update_file_in_vector_db(
    state: State<'_, AppState>,
    file_path: String,
) -> Result<(), String> {
    #[cfg(feature = "vector_db")]
    {
        let db_guard = state.vector_db.lock().await;
        if let Some(ref db) = *db_guard {
            match db.update_file(&file_path).await {
                Ok(_) => {
                    info!("Updated file '{}' in vector database", file_path);
                    Ok(())
                }
                Err(e) => Err(format!("Failed to update file in vector database: {}", e)),
            }
        } else {
            Err("Vector database not initialized".to_string())
        }
    }
    #[cfg(not(feature = "vector_db"))]
    {
        Err("Vector database feature not enabled".to_string())
    }
}

#[tauri::command]
pub async fn get_indexed_files(state: State<'_, AppState>) -> Result<Vec<String>, String> {
    #[cfg(feature = "vector_db")]
    {
        let db_guard = state.vector_db.lock().await;
        if let Some(ref db) = *db_guard {
            let files = db.get_indexed_files().await;
            Ok(files)
        } else {
            Err("Vector database not initialized".to_string())
        }
    }
    #[cfg(not(feature = "vector_db"))]
    {
        Err("Vector database feature not enabled".to_string())
    }
}

#[tauri::command]
pub async fn test_vector_db_connection(state: State<'_, AppState>) -> Result<(), String> {
    #[cfg(feature = "vector_db")]
    {
        let db_guard = state.vector_db.lock().await;
        if let Some(ref db) = *db_guard {
            match db.test_connection().await {
                Ok(_) => {
                    info!("Vector database connection test successful");
                    Ok(())
                }
                Err(e) => Err(format!("Vector database connection test failed: {}", e)),
            }
        } else {
            Err("Vector database not initialized".to_string())
        }
    }
    #[cfg(not(feature = "vector_db"))]
    {
        Err("Vector database feature not enabled".to_string())
    }
}

#[tauri::command]
pub async fn get_embedding_dimension(state: State<'_, AppState>) -> Result<usize, String> {
    #[cfg(feature = "vector_db")]
    {
        let db_guard = state.vector_db.lock().await;
        if let Some(ref db) = *db_guard {
            match db.get_embedding_dimension().await {
                Ok(dimension) => Ok(dimension),
                Err(e) => Err(format!("Failed to get embedding dimension: {}", e)),
            }
        } else {
            Err("Vector database not initialized".to_string())
        }
    }
    #[cfg(not(feature = "vector_db"))]
    {
        Err("Vector database feature not enabled".to_string())
    }
}

File: src/types/vectordb.ts
// TypeScript interfaces for Vector Database Tauri commands

export interface SearchResultResponse {
  id: string;
  score: number;
  payload: Record<string, any>;
}

export interface VectorDbStats {
  collection_info: any;
  point_count: number;
  mapped_files: number;
  embedding_model: string;
}

export interface AddFileResponse {
  point_id: string;
  file_path: string;
}

export interface SyncFilesResponse {
  added_files: string[];
  count: number;
}

export interface VectorDbCommands {
  /**
   * Create a new vector database with the given collection name
   */
  create_vector_db: (name: string) => Promise<void>;

  /**
   * Delete the vector database and all its data
   */
  delete_vector_db: () => Promise<void>;

  /**
   * Add a file to the vector database
   */
  add_file_to_vector_db: (file_path: string) => Promise<AddFileResponse>;

  /**
   * Search the vector database for similar content
   */
  search_vector_db: (
    query: string,
    top_k?: number,
  ) => Promise<SearchResultResponse[]>;

  /**
   * Search the vector database with additional filters
   */
  search_vector_db_with_filter: (
    query: string,
    top_k: number | undefined,
    filter: Record<string, any>,
  ) => Promise<SearchResultResponse[]>;

  /**
   * Delete a file from the vector database by its path
   */
  delete_file_from_vector_db: (file_path: string) => Promise<void>;

  /**
   * Delete a specific point from the vector database by its ID
   */
  delete_point_from_vector_db: (point_id: string) => Promise<void>;

  /**
   * Delete all embeddings from the vector database
   */
  delete_all_embeddings: () => Promise<void>;

  /**
   * Get statistics about the vector database
   */
  get_vector_db_stats: () => Promise<VectorDbStats>;

  /**
   * Sync files from a directory to the vector database
   */
  sync_files_to_vector_db: (
    directory_path: string,
  ) => Promise<SyncFilesResponse>;

  /**
   * Update an existing file in the vector database
   */
  update_file_in_vector_db: (file_path: string) => Promise<void>;

  /**
   * Get a list of all indexed files in the vector database
   */
  get_indexed_files: () => Promise<string[]>;

  /**
   * Test the vector database connection
   */
  test_vector_db_connection: () => Promise<void>;

  /**
   * Get the embedding dimension used by the model
   */
  get_embedding_dimension: () => Promise<number>;
}

// Helper type for invoke calls
export type VectorDbCommandName = keyof VectorDbCommands;

// Search options interface for more complex queries
export interface VectorSearchOptions {
  query: string;
  topK?: number;
  scoreThreshold?: number;
  filter?: {
    extension?: string;
    fileSize?: { min?: number; max?: number };
    modified?: { after?: number; before?: number };
    filePath?: string;
  };
}

// File metadata interface
export interface FileMetadata {
  file_path: string;
  file_name: string;
  file_size?: number;
  modified?: number;
  extension?: string;
  content_preview?: string;
}

// Enhanced search result with typed payload
export interface TypedSearchResult {
  id: string;
  score: number;
  metadata: FileMetadata;
}

// Vector database status
export enum VectorDbStatus {
  NotInitialized = "not_initialized",
  Initializing = "initializing",
  Ready = "ready",
  Error = "error",
}

// Configuration interfaces
export interface EmbedConfig {
  model_name: string;
  max_length: number;
  batch_size: number;
  show_download_progress: boolean;
  cache_dir?: string;
}

export interface QdrantConfig {
  enabled: boolean;
  server_url: string;
  server_port: number;
  collection_name: string;
  vector_size: number;
  distance_metric: string;
  timeout_seconds: number;
  use_grpc: boolean;
}

export interface VectorDbConfig {
  qdrant_config: QdrantConfig;
  embed_config: EmbedConfig;
}

// Error types
export class VectorDbError extends Error {
  constructor(
    message: string,
    public readonly code?: string,
    public readonly details?: any,
  ) {
    super(message);
    this.name = "VectorDbError";
  }
}

// Utility functions for working with search results
export class VectorDbUtils {
  /**
   * Convert search result to typed format
   */
  static toTypedResult(result: SearchResultResponse): TypedSearchResult {
    return {
      id: result.id,
      score: result.score,
      metadata: {
        file_path: result.payload.file_path as string,
        file_name: result.payload.file_name as string,
        file_size: result.payload.file_size as number,
        modified: result.payload.modified as number,
        extension: result.payload.extension as string,
        content_preview: result.payload.content_preview as string,
      },
    };
  }

  /**
   * Filter results by minimum score
   */
  static filterByScore(
    results: SearchResultResponse[],
    minScore: number,
  ): SearchResultResponse[] {
    return results.filter((result) => result.score >= minScore);
  }

  /**
   * Group results by file extension
   */
  static groupByExtension(
    results: TypedSearchResult[],
  ): Record<string, TypedSearchResult[]> {
    return results.reduce(
      (groups, result) => {
        const ext = result.metadata.extension || "unknown";
        if (!groups[ext]) {
          groups[ext] = [];
        }
        groups[ext].push(result);
        return groups;
      },
      {} as Record<string, TypedSearchResult[]>,
    );
  }

  /**
   * Sort results by relevance score (descending)
   */
  static sortByRelevance(
    results: SearchResultResponse[],
  ): SearchResultResponse[] {
    return [...results].sort((a, b) => b.score - a.score);
  }

  /**
   * Extract unique file paths from results
   */
  static extractFilePaths(results: SearchResultResponse[]): string[] {
    return [...new Set(results.map((r) => r.payload.file_path as string))];
  }
}

// React hook type for vector database operations
export interface UseVectorDbReturn {
  // State
  isInitialized: boolean;
  isLoading: boolean;
  error: string | null;
  stats: VectorDbStats | null;

  // Core operations
  search: (
    query: string,
    options?: Partial<VectorSearchOptions>,
  ) => Promise<SearchResultResponse[]>;
  addFile: (filePath: string) => Promise<AddFileResponse>;
  deleteFile: (filePath: string) => Promise<void>;
  syncDirectory: (directoryPath: string) => Promise<SyncFilesResponse>;
  refreshStats: () => Promise<void>;
  testConnection: () => Promise<boolean>;

  // Extended operations
  initializeDb: (collectionName?: string) => Promise<boolean>;
  updateFile: (filePath: string) => Promise<void>;
  getIndexedFiles: () => Promise<string[]>;
  clearDatabase: () => Promise<void>;
  getEmbeddingDimension: () => Promise<number>;
  clearError: () => void;
  checkInitialization: () => Promise<boolean>;
}

export default VectorDbCommands;
